사칙연산 PR 입니당

limwoobin opened the pull request:
racingcar 랑 따로 pr 올리려구여 ㅎㅎ

limwoobin committed:
Add - 계산기 테스트코드 추가 , 계산기 클래스 및 enum 타입 추가

limwoobin committed:
Add - validationCheck 테스트 코드 추가

limwoobin committed:
Fix - Validator Class 추가에 의한 테스트코드 수정 , 사칙연산 테스트메소드 추가(구현 x)

limwoobin committed:
Add - Validator 객체 추가

limwoobin committed:
Change - Symbol.java 파일 이름 변경 (Operator.java)

limwoobin committed:
Delete - ExpressionOrder.java 파일 삭제 (사용 안함)

limwoobin committed:
Fix - 사칙연산 테스트코드 추가 및 서비스 코드 사칙연산 메소드 분리 (각 메소드당 하나의 일을 하는 방향으로 분리)

limwoobin committed:
FIx - 불필요 Test Code 삭제

limwoobin committed:
Fix - Operator 변수명 변경

limwoobin committed:
Pr Test

limwoobin committed:
Pr Test2

limwoobin committed:
Pr Test3

updown-kurly commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,72 @@
+
+    public int calculate(String expression) {
+        validator.validationCheck(expression);
+        for (String data : expression.split(" ")) {
```
" " - > BLACK 같은 상수로 처리한다면 좋을것 같습니다

limwoobin replied:
넵 문자열을 상수로 처리하는 습관을 들여야겠습니다

updown-kurly commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,72 @@
+    private void updateOperand(String data) {
+        if (!isNumber(data)) {
+            operator = data;
+        } else {
```
객체지향 생활 체조원칙을 지켜보는건 어떨까요?
규칙 2: else 예약어를 쓰지 않는다.

limwoobin replied:
좋은의견 감사합니다!

updown-kurly commented on src/main/java/calculator/Operator.java:
```diff
@@ -0,0 +1,26 @@
+    DIVISION("/" , (previousValue , nextValue) -> previousValue / nextValue);
+
+    private String operator;
+    private BiFunction<Integer , Integer , Integer> calculate;
```
👍

updown-kurly commented on src/main/java/calculator/ExpressionValidator.java:
```diff
@@ -0,0 +1,27 @@
+
+    public Operator toEnum(String operator) {
+        for (Operator operationOperator : Operator.values()) {
+            if (operationOperator.getOperator().equals(operator)) {
+                return operationOperator;
+            }
```
객체지향 생활 체조원칙을 지켜보는건 어떨까요?
규칙 1: 한 메서드에 오직 한 단계의 들여쓰기만 한다.

limwoobin replied:
좋은의견 감사합니다!

updown-kurly commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,72 @@
+        this.nextValue = 0;
+        this.result = 0;
+        this.operator = null;
+        this.validator = new ExpressionValidator();
```
유틸성으로 사용할 클래스라면 사용해야할 클래스마다 인스턴스 생성하지않고 static 으로 올려서
ExpressionValidator. validationCheck(expressiont); 이렇게 가는게 더 좋을것 같습니다!

limwoobin replied:
좋은의견 감사합니다!
지금보니 static 으로 사용하는게 더 알맞는것같네요

updown-kurly commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,72 @@
+    private final ExpressionValidator validator;
+
+    public Calculator() {
+        this.previousValue = 0;
+        this.nextValue = 0;
+        this.result = 0;
+        this.operator = null;
+        this.validator = new ExpressionValidator();
```
생성자에서 이렇게 초기화 되어야하는 이유가 무엇인가용??

limwoobin replied:
this.validator = new ExpressionValidator(); 
는 말씀해주신 static 으로 변경하는게 더 알맞아보입니다.

updown-kurly commented on src/test/java/calculator/CalculatorTest.java:
```diff
@@ -0,0 +1,126 @@
+            String 널_VALUE = null;
+            String 빈_공백 = " ";
+
+            RuntimeException 널값을_넘긴_경우 = assertThrows(
+                    RuntimeException.class,
+                    () -> calculator.calculate(널_VALUE));
+
+            assertThat(널값을_넘긴_경우.getClass()).isEqualTo(IllegalArgumentException.class);
```
assertThatExceptionOfType(IllegalArgumentException.class).isThrownBy(() -> {
   calculator.calculate(널_VALUE);
});
로 하면 되나요?

updown-kurly commented on src/main/java/calculator/Operator.java:
```diff
@@ -0,0 +1,26 @@
+
+public enum Operator {
+    PLUS("+" , Integer::sum),
+    MINUS("-"  , (previousValue , nextValue) -> previousValue - nextValue),
```
얘만 "-"공백공백 두개 들어가 있네요 
cmd+shift+enter
나 포멧팅 단축키 사용한다면 전부 동일하게 통일 될 것 같습니다.

limwoobin replied:
꿀팁 감사합니다.

updown-kurly reviewed (commented):
간단한 리뷰사항 드렸습니다

eeunheechoi commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,72 @@
+        this.previousValue = 0;
+        this.nextValue = 0;
+        this.result = 0;
+        this.operator = null;
```
입력 받은 식을 차례로 연산하면 되니 for문을 수행하여 처음 선언한 변수를 재활용 하여 연산을 해나가면 변수 선언을 줄일 수 있을거 같아요. 아래는 제가  피드백 받고 구현한 로직이에요. 참고해 보세요.
formula는 입력받은 식을 " "으로 split하여 배열에 넣은 것입니다. 
![image](https://user-images.githubusercontent.com/46669136/114475323-fd7c1f00-9c32-11eb-8e9b-8a5a6e154881.png)

limwoobin replied:
좋은의견 감사합니다!
한번 참고하겠습니다 ㅎㅎ

eeunheechoi commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,72 @@
+        this.previousValue = result;
+        this.operator = null;
+        this.nextValue = 0;
+    }
```
Calculator가 전반적으로 하는일이 많아 보여요. 계산과 식을 분리해서 클래스 작성해도 좋을 거 같아요

limwoobin replied:
Calculator와 ExpressionValidator 로 분리해서
최대한 객체가 SRP 를 지킬수있게 노력해봐야겠네요 ㅎㅎ

eeunheechoi reviewed (approved):
저는 처음 계산기 짰을때 엉망이었는데 enum사용 하신거 보고 놀랐어요 멋지네요. 저도 많이 가르쳐 주세요.

hyerinhong commented on src/main/java/calculator/requirements.md:
```diff
@@ -0,0 +1,20 @@
+## 기능 요구사항
+
+- 사용자가 입력한 문자열 값에 따라 사칙연산을 수행할 수 있는 계산기를 구현해야 한다.
+- 입력 문자열의 숫자와 사칙 연산 사이에는 반드시 빈 공백 문자열이 있다고 가정한다.
+- 나눗셈의 경우 결과 값을 정수로 떨어지는 값으로 한정한다.
+- 문자열 계산기는 사칙연산의 계산 우선순위가 아닌 입력 값에 따라 계산 순서가 결정된다. 즉, 수학에서는 곱셈, 나눗셈이 덧셈, 뺄셈 보다 먼저 계산해야 하지만 이를 무시한다.
+- 예를 들어 2 + 3 * 4 / 2와 같은 문자열을 입력할 경우 2 + 3 * 4 / 2 실행 결과인 10을 출력해야 한다.
```
기능 요구사항을 쓸 때, 제시된 제약사항을 그대로 적는 것도 좋지만 
하나의 요구사항이 하나의 테스트케이스를 구성한다고 생각하고 작성해봐도 좋을거 같아요 

- 입력 문자열의 숫자와 사칙 연산 사이에는 반드시 빈 공백 문자열이 있다고 가정한다 
-- 빈 공백 문자열이 2개일 경우, Exception 이 발생해야한다 
-- 숫자와 사칙연산 사이에 공백이 아닌 문자가 있다면 Exception이 발생해야한다 

이런 식으로 기능 요구사항을 좀 더 디테일하게 작성하게 되면 테스트를 작성해야하는 케이스들이 나열되게 되고, 
TDD 로 좀 더 잘 개발할 수 있는 거 같아요

limwoobin replied:
좋은 꿀팁 감사합니다!

hyerinhong commented on src/main/java/calculator/Operator.java:
```diff
@@ -0,0 +1,26 @@
+    PLUS("+" , Integer::sum),
+    MINUS("-"  , (previousValue , nextValue) -> previousValue - nextValue),
+    MULTIPLY("*" , (previousValue , nextValue) -> previousValue * nextValue),
+    DIVISION("/" , (previousValue , nextValue) -> previousValue / nextValue);
```
0으로 나누면 어떻게 될까요? 
이 부분에 대한 예외처리도 추가해볼까요? :)

limwoobin replied:
넵 계산식에 대한 에외발생 케이스를 전혀 고려하지 않았네요
테스트 코드에도 누락되어서 좀 더 세밀한 테스트케이스를 작성해야 할 것 같습니다.

hyerinhong commented on src/main/java/calculator/ExpressionValidator.java:
```diff
@@ -0,0 +1,27 @@
+
+    public Operator toEnum(String operator) {
+        for (Operator operationOperator : Operator.values()) {
+            if (operationOperator.getOperator().equals(operator)) {
```
toEnum은 validator에 있는 거 보다, enum이 가지는게 맞아보여요 

getOperator().equals 와 같이 get한 다음 밖에서 비교하지 말고,
Operator enum에게 물어보면 어떨까요? 이 symbol에 해당하는 Enum이 무엇인지

limwoobin replied:
넵 toEnum 은 enum 으로 전환하는거니 
enum 객체에게 권한을 위임하라는 뜻으로 이해하면 될까요?

hyerinhong commented on src/main/java/calculator/Calculator.java:
```diff
@@ -0,0 +1,72 @@
+        if (!isNumber(data)) {
+            operator = data;
+        } else {
+            if (previousValue == 0 && nextValue == 0) {
+                this.previousValue = toInt(data);
+            } else {
+                this.nextValue = toInt(data);
+            }
```
0 - 1 은 계산 결과가 어떻게 나오나요?

limwoobin replied:
이 부분도 위에서 코멘트남겨주신거와 마찬가지로
좀 더 세밀한 테스트 코드를 짜야겠습니다 
리뷰 감사합니다ㅣ

hyerinhong reviewed (commented):
간단하게 저도 코멘트 몇개 드렸습니다 :) 

프로그램에 main 이 안보이네요?
테스트케이스를 통해 기본로직 검증이 가능하지만, console에서 다양하게 실행도 가능하도록 구현해보면 좋을 거 같습니다

limwoobin committed:
Fix - PR review 반영

1. 요구사항 세분화
2. 나누기의 경우 0에 대한 exception code 추가
3. util 성격의 메소드 static 으로 변경
4. 예외사항에 대한 TestCode 추가

limwoobin committed:
Fix - PR 추가 review 반영

계산 방식 수정

hhlyn0328 commented on src/main/java/calculator/ExpressionValidator.java:
```diff
@@ -0,0 +1,17 @@
+    }
+
+    private static void inputDataRegex(String data) {
+        if (data == null || "".equals(data.trim())) {
```
동일한 역할을 하는 method 가 있습니다 
StringUtils.isBlank 를 한 번 써보세요 :)

hhlyn0328 reviewed (approved):
고생하셨습니다 :) 
마지막 코멘트가 늦어서 죄송해요 머지할게요 
코멘트 1개 추가한건 참고만 해주세요 

이어서 자동차 미션도 진행해주십쇼 !
